protocols/relay/src/handler.rs

```diff
@@ -0,0 +1,755 @@
+                                }
+                                // With either status below there is no reason to stay connected.
+                                // Thus terminate the connection.
+                                circuit_relay::Status::HopSrcAddrTooLong
+                                | circuit_relay::Status::HopSrcMultiaddrInvalid
+                                | circuit_relay::Status::MalformedMessage => {
+                                    self.pending_error = Some(ProtocolsHandlerUpgrErr::Upgrade(
+                                        upgrade::UpgradeError::Apply(EitherError::B(
+                                            EitherError::A(error),
+                                        )),
+                                    ));
+                                }
+                                // While useless for reaching this particular destination, the
+                                // connection to the relay might still proof helpful for other
+                                // destinations. Thus do not terminate the connection.
+                                circuit_relay::Status::StopSrcAddrTooLong
+                                | circuit_relay::Status::StopDstAddrTooLong
+                                | circuit_relay::Status::StopSrcMultiaddrInvalid
+                                | circuit_relay::Status::StopDstMultiaddrInvalid
+                                | circuit_relay::Status::StopRelayRefused
+                                | circuit_relay::Status::HopDstAddrTooLong
+                                | circuit_relay::Status::HopDstMultiaddrInvalid
+                                | circuit_relay::Status::HopNoConnToDst
+                                | circuit_relay::Status::HopCantDialDst
+                                | circuit_relay::Status::HopCantOpenDstStream
+                                | circuit_relay::Status::HopCantSpeakRelay
+                                | circuit_relay::Status::HopCantRelayToSelf => {}
+                            }
+                        }
+                    },
+                    ProtocolsHandlerUpgrErr::Upgrade(upgrade::UpgradeError::Apply(
+                        EitherError::B(_),
+                    )) => {
+                        unreachable!("Can not receive an OutgoingDstReqError when dialing a relay.")
+                    }
+                }
+
+                self.queued_events
+                    .push(RelayHandlerEvent::OutgoingRelayReqError(
+                        dst_peer_id,
+                        request_id,
+                    ));
+            }
+            RelayOutboundOpenInfo::Destination {
+                src_connection_id,
+                incoming_relay_req,
+                ..
+            } => {
+                let err_code = match error {
+                    ProtocolsHandlerUpgrErr::Timeout => {
+                        self.pending_error = Some(ProtocolsHandlerUpgrErr::Timeout);
```

Comment:
Do we really want to terminate the connection on a single relay substream negotiation timeout? Other protocols may use the connection (successfully) as well, right?

Reply:
If I understand correctly, given that `libp2p-relay` uses the default timeout, this will only happen after 10 seconds. I would deem a node that is not able to respond to the light-weight circuit relay requests within 10 seconds as either misbehaving or overloaded. For the former, disconnecting seems to be the way to go. For the latter I would say disconnecting is beneficial for both sides. The local node might be able to succeed in whatever it is up to via another relay or destination node. In addition I would guess other protocols running on the same connection are not making progress either if the light-weight circuit relay negotiation does not make progress. The overloaded remote node would receive less traffic (both through the relay protocol and any other protocol out there) and thus become less overloaded overall

The above said, I don't know what the correct behaviour is and I don't know whether it should be consistent across protocols. E.g. `libp2p-request-response` does not terminate the whole connection on `Timeout`, which makes sense to me as the payloads being exchanged could potentially be very large.

https://github.com/libp2p/rust-libp2p/blob/f48bb15fa7749187570d4e092dbc2346cee188a0/protocols/request-response/src/handler.rs#L252-L260

With my reasoning above in mind, what do you think @romanb we should do in case of a `ProtocolsHandlerUpgrErr::Timeout`?